peer/sync.go

```diff
@@ -0,0 +1,152 @@
+// content will be awaken and will create their own copies of the content.
+func (pending *pendingContent) signal(content []byte) {
+	pending.cond.L.Lock()
+	pending.content = content
+	pending.cond.L.Unlock()
+}
+
+type Syncer struct {
+	opts      SyncerOptions
+	filter    *channel.SyncFilter
+	transport *transport.Transport
+
+	pendingMu *sync.Mutex
+	pending   map[string]pendingContent
+}
+
+func NewSyncer(opts SyncerOptions, filter *channel.SyncFilter, transport *transport.Transport) *Syncer {
+	return &Syncer{
+		opts:      opts,
+		filter:    filter,
+		transport: transport,
+
+		pendingMu: new(sync.Mutex),
+		pending:   make(map[string]pendingContent, 1024),
+	}
+}
+
+func (syncer *Syncer) Sync(ctx context.Context, contentID []byte, hint *id.Signatory) ([]byte, error) {
+	syncer.pendingMu.Lock()
+	pending, ok := syncer.pending[string(contentID)]
+	if !ok {
+		pending = pendingContent{
+			content: nil,
+			cond:    sync.NewCond(new(sync.Mutex)),
+		}
+		syncer.pending[string(contentID)] = pending
+	}
+	syncer.pendingMu.Unlock()
+
+	// Allow synchronisation messages for the content ID. This is required in
+	// order for channel to not filter inbound content (of unknown size). At the
+	// end of the method, we Deny the content ID again, un-doing the Allow and
+	// blocking content again.
+	syncer.filter.Allow(contentID)
+	defer syncer.filter.Deny(contentID)
+
+	if ok {
+		select {
+		case <-ctx.Done():
+			return nil, ctx.Err()
+		case content := <-pending.wait():
+			return content, nil
+		}
+	}
+
+	// Get addresses close to our address. We will iterate over these addresses
+	// in order and attempt to synchronise content by sending them pull
+	// messages.
+	peers := syncer.transport.Table().Peers(syncer.opts.Alpha)
+	if hint != nil {
+		peers = append([]id.Signatory{*hint}, peers...)
+	}
+
+	for _, peer := range peers {
+		content, err := func() ([]byte, error) {
+			ctx, cancel := context.WithTimeout(ctx, syncer.opts.Timeout)
+			defer cancel()
+
+			err := syncer.transport.Send(ctx, peer, wire.Msg{
+				Version: wire.MsgVersion1,
+				Type:    wire.MsgTypePull,
+				Data:    contentID,
+			})
+			if err != nil {
+				return nil, fmt.Errorf("pulling: %v", err)
+			}
+
+			select {
+			case <-ctx.Done():
+				return nil, ctx.Err()
+			case content := <-pending.wait():
+				return content, nil
+			}
+		}()
+		if err != nil {
+			syncer.opts.Logger.Debug("sync", zap.String("peer", peer.String()), zap.Error(err))
+			continue
+		}
+		return content, nil
+	}
+
+	return nil, fmt.Errorf("content not found: %v", base64.RawURLEncoding.EncodeToString(contentID))
+}
+
+func (syncer *Syncer) DidReceiveMessage(from id.Signatory, msg wire.Msg) error {
+	if msg.Type == wire.MsgTypeSync {
+		if !syncer.filter.Filter(from, msg) {
+			return fmt.Errorf("denied message from %v", from)
+		}
+		syncer.pendingMu.Lock()
+		pending, ok := syncer.pending[string(msg.Data)]
+		if ok && msg.SyncData != nil {
```

Comment:
This will always go through once a `contentID` is registered as pending (even if the message has already been synced). Maybe we should delete the corresponding `pendingContent` from the map before returning from `Sync`?

Reply:
This is fine, because the SyncFilter being used will prevent Sync messages (see the calls to Allow and Deny). However, we should delete the entry from the map to prevent unbounded growth, but that is a different problem.

Reply:
Ah yep, the filter should weed out unwanted messages. An issue with deleting the it from the map is that if multiple goroutines are calling `Sync` at the same time, it will introduce a race condition